h5topng.1
.\" Copyright (c) 1999, 2000 Massachusetts Institute of Technology
.\" 
.\" Permission is hereby granted, free of charge, to any person obtaining
.\" a copy of this software and associated documentation files (the
.\" "Software"), to deal in the Software without restriction, including
.\" without limitation the rights to use, copy, modify, merge, publish,
.\" distribute, sublicense, and/or sell copies of the Software, and to
.\" permit persons to whom the Software is furnished to do so, subject to
.\" the following conditions:
.\" 
.\" The above copyright notice and this permission notice shall be
.\" included in all copies or substantial portions of the Software.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
.\" EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
.\" CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
.\" TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
.\" SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
.\"
.TH H5TOPNG 1 "November 21, 1999" "h5utils" "h5utils"
.SH NAME
h5topng \- generate PNG images from 2d slices of HDF5 files
.SH SYNOPSIS
.B h5topng
[\fIOPTION\fR]... [\fIHDF5FILE\fR]...
.SH DESCRIPTION
.PP
." Add any additional description here
h5topng is a utility to generate images in PNG (Portable Network Graphics)
format from two-dimensional slices of datasets in HDF5 files.  It is
designed for quick-and-dirty visualization of scientific data, and for
batch processing thereof via shell scripts.

HDF5 is a free, portable binary format and supporting library developed
by the National Center for Supercomputing Applications at the University
of Illinois in Urbana-Champaign.  A single
.I h5
file can contain multiple data sets; by default,
.I h5topng
takes the first dataset, but this can be changed via the
.B -d
option, or by using the syntax \fIHDF5FILE:DATASET\fR.

For a three-dimensional data set, the xy slice at z index zero is extracted
by default, but other slices can be extracted using the
.B -xyz
options.  Yet more options control things like the colormap and
magnification.  Still, the most basic usage is something like
\'h5topng foo.h5\', which will output a file foo.png containing an image
from the data in foo.h5.
.SH OPTIONS
.TP
.B -h
Display help on the command-line options and usage.
.TP
.B -V
Print the version number and copyright info for h5topng.
.TP
.B -v
Verbose output.  This output includes the minimum and maximum values
encountered in the data, which is useful to know for the
.B -mM
options.
.TP
\fB\-o\fR \fIfile\fR
Send PNG output to
.I file
rather than to the filename with .h5 replaced with .png (the default).
.TP
\fB\-x\fR \fIix\fR, \fB\-y\fR \fIiy\fR, \fB\-z\fR \fIiz\fR
This tells
.I h5topng
to use a particular slice of a two- or three-dimensional dataset.  e.g.
.B -x
causes a yz plane to be used at an x index of
.I ix
(where the indices run from zero to one less than the maximum index in
that direction).  The default behaviour is equivalent to
.B -z
0.
.TP
\fB\-X\fR \fIscalex\fR, \fB\-Y\fR \fIscaley\fR
Scale the x and y dimensions by
.I scalex
and
.I scaley
respectively.  The default is to use scale factors of 1.0;
i.e. the image has the same dimensions (in pixels) as the data.  Linear
interpolation is used to fill in the pixels when the scale factors are
not 1.0.
.TP
\fB\-s\fR \fIskewangle\fR
Skew the image by
.I skewangle
(in degrees) to the left or right.  The result is a parallelogram, with
the leftover space in the (square) image filled with either black or white
pixels, depending upon the color map.
.TP
.B -T
Transpose the data (interchange the image axes).  By default, no
transposition is done.
.B -c
Use a color map going from blue to white to red, with white in the center,
instead of the default color map which is a grayscale ramp from white to
black.  If you use the
.B -c
option, you may also want to use the
.B -mM
options so that the center of the color scale (white) corresponds to
something like zero (if your data consists of positive and negative
values).  In this case, you would want to use values of equal magnitude
and opposite sign for the parameters to the
.B -m
and 
.B -M
options.
.TP -r
Reverse the ordering of the color map.
.TP
\fB\-m\fR \fImin\fR, \fB\-M\fR \fImax\fR
Normally, the bottom and top of the color map correspond to the
minimum and maximum values in the data.  Using these options, you
can make the bottom and top of the color map correspond to
.I min
and
.I max
instead.  Data values below or above this range will be treated as if
they were
.I min
or
.I max
respectively.
.TP
\fB\-C\fR \fIfile\fR, \fB\-b\fR \fIval\fR
Superimpose contour outlines from the first dataset in the
.I file
HDF5 file, which should have the same dimensions as the input
dataset, on all of the output images.  The contour outlines are around
values that differ from
.I val
(default 1.0).
.TP
\fB\-d\fR \fIname\fR
Use dataset
.I name
from the input files; otherwise, the first dataset from each file is used.
Alternatively, use the syntax \fIHDF5FILE:DATASET\fR, which allows you
to specify a different dataset for each file.
You can use the
.I h5ls
command (included with hdf5) to find the names of datasets within a file.
.SH BUGS
Send bug reports to S. G. Johnson, stevenj@alum.mit.edu.
.SH AUTHORS
Written by Steven G. Johnson.  Copyright (c) 1999, 2000 by the Massachusetts
Institute of Technology.


NEWS
h5utils 1.10.1 (9/20/06)

  * Fixed build problem on Cygwin due to missing ".exe" extension. 
    Thanks to Ken Hill for the bug report.

h5utils 1.10 (9/2/05)

  * Added h4fromh5 utility.  (NCSA seems to be no longer shipping
    the HDF5/HDF4 conversion tools with the latest HDF5 release.)

  * Added dkbluered color map, which is similar to bluered but
    uses a somewhat wider range of colors.

h5utils 1.9.1 (8/5/04)

  * Fix h5topng compilation failure with some non-C99 compilers;
    thanks to Maarten van Reeuwijk for the bug report.

h5utils 1.9 (7/12/04)

  * Added new h5math utility, which creates and combines HDF5 datasets
    using a user-specified mathematical expression.  (Requires GNU
    libmatheval to be installed.)

  * h5topng: Matlab-like start:end or start:step:end notation for
    slice indices, to allow a whole sequence of slices to be output as
    a sequence of PNG images.

  * h5topng: if contour/overlay dataset does not have same dimensions
    as output data, it is periodically "tiled" over the output.

h5utils 1.8 (5/22/04)

  * New -A and -a options for h5topng to allow translucent overlays
    from one file onto another (an alternative or complement to the
    -C contour-overlay option).

  * h5topng uses 24-bit direct color by default (use -8 option for
    old 8-bit behavior).

  * h5topng uses columns/rows for x/y by default (use -T to swap),
    the opposite of the old behavior.

  * There is no default -z 0 slice dimension in h5topng/h5totxt any more.
    You must specify a slice for 3+ dimensional data in h5topng.
    h5totxt dumps the whole data file by default unless one or more
    slices are specified.

  * Support specifying multiple slice dimensions for 4+ dimensional
    datasets, with new "-t" option to indicate final dimension.

  * Slices are also supported now in h5tovtk and h5tov5d.

  * New -. option in h5totxt to specify number of significant digits; 
    output 16 digits by default instead of 6, previously.

h5utils 1.7.1 (3/16/02)

  * Fixed array overrun in h5topng that caused a floating-point exception
    on Alphas; thanks to Marin Soljacic for the bug report.

h5utils 1.7 (3/9/02)

  * h5topng now supports multiple, user-definable color tables, a number
    of which are provided.  INCOMPATIBLE CHANGE: the -c option now has
    the syntax: -c <colortable>.  The old behavior corresponds to the
    included "bluered" colortable, invoked via: -c bluered.

  * New -R option for h5topng to use a consistent color scale for
    all specified files.

  * New -0 option for h5topng and h5totxt that shifts the origin of the
    slice coordinates to the dataset center.

  * Added h5tovtk program to output VTK (Visualization ToolKit) data files.

  * Support -T (transpose dimensions) option in h5tov5d.

  * Fixed bug in h5topng that caused extra rows/columns of garbage pixels
    to be written at the edges of images when scaling was used.

  * When compiling the h5read Octave plugin, respect the CPPFLAGS and LDFLAGS 
    environment variables.  Thanks to Max Colice for the bug report.

  * Fixed problem when --without-h5tov5d and --without-h5fromh4 are used. 
    Thanks to Nikola Ivanov Nikolov for the fix.

h5utils 1.6 (1/17/01)

  * Don't build h5fromh4 if the superior h4toh5 tool (from HDF5 1.4)
    is present.  Also added --{with,without}-h5fromh4 option to
    configure to force whether h5fromh4 is built.

h5utils 1.5.1 (12/9/00)

  * Support manually disabling Octave plugin support
    (configure --without-octave) in case of C++ problems.

  * Support Vis5d+ and Debian HDF header file locations.

h5utils 1.5 (7/9/00)

  * Added h5fromh4 program to convert HDF4 datasets to HDF5 format.

  * Added -S <s> option to h5topng as a shortcut for -X <s> -Y <s>.

h5utils 1.4 (5/28/00)

  * Added h5fromtxt program to convert text input to an HDF5 dataset.

h5utils 1.3.4 (1/31/00)

  * Improved -C contour plotting in h5topng.

  * Fix in h5topng man page (thanks to Christoph Becher).

h5utils 1.3.3 (1/30/00)

  * Bug fix in h5topng (would sometimes output solid black images).
    Thanks to Karl Koch for the bug report.

h5utils 1.3.2 (1/28/00)

  * Added h5topng -Z option to center color scale on zero.

  * Now support h5topng -C filename:dataset.

h5utils 1.3.1 (1/27/00)

  * Bug fixes in dataset name-handling, especially when using h5tov5d to
    join multiple datasets into one output file.

h5utils 1.3 (1/21/00)

  * You can now specify individual datasets within a file by using
    <filename>:<dataset> instead of just <filename> with h5topng,
    h5totxt, and h5tov5d.

h5utils 1.2.3: (1/20/00)

  * Fixed minor bug in 'h5totxt -h'.

h5utils 1.2.2: (1/12/00)

  * Makefile now includes CPPFLAGS in the compiler flags, making it
    easier to use header files in non-standard locations.  (CPPFLAGS
    is the proper place to put -I flags for the configure script.)

h5utils 1.2.1:  (1/1/00)

  * Modified -o option of h5tov5d to join datasets into a single Vis5d file.

h5utils 1.2:  (12/31/99)

  * Added h5tov5d program for converting to Vis5d format.

  * Improved installation; h5read.oct now goes into the sitewide Octave
    plugins directory, and things work correctly when only some of the
    utilities are compiled.

h5utils 1.1:  (12/6/99)

  * Added h5totxt program for exporting 2d slices of HDF5 files to text
    suitable for importing into a spreadsheet.

h5utils 1.0:  (11/22/99)

  * Initial release.


h5topng.c
/* Copyright (c) 1999, 2000 Massachusetts Institute of Technology
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 * 
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <unistd.h>

#include "config.h"
#include "arrayh5.h"
#include "copyright.h"
#include "writepng.h"

#define CHECK(cond, msg) { if (!(cond)) { fprintf(stderr, "h5topng error: %s\n", msg); exit(EXIT_FAILURE); } }

void usage(FILE *f)
{
     fprintf(f, "Usage: h5topng [options] [<filenames>]\n"
	     "Options:\n"
	     "         -h : this help message\n"
             "         -V : print version number and copyright\n"
	     "         -v : verbose output\n"
	     "  -o <file> : output to <file> (first input file only)\n"
	     "    -x <ix> : take x=<ix> slice of data\n"
	     "    -y <iy> : take y=<iy> slice of data\n"
	     "    -z <iz> : take z=<iz> slice of data [ default: z=0 ]\n"
	     "    -X <sx> : scale width by <sx> [ default: 1.0 ]\n"
	     "    -Y <sy> : scale height by <sy> [ default: 1.0 ]\n"
	     "  -s <skew> : skew axes by <skew> degrees [ default: 0 ]\n"
	     "         -T : transpose the data [default: no]\n"
	     "         -c : blue-white-red palette instead of grayscale\n"
	     "         -r : reverse color map [default: no]\n"
	     "   -m <min> : set bottom of color scale to data value <min>\n"
	     "   -M <max> : set top of color scale to data value <max>\n"
	     "  -C <file> : superimpose contour outlines from <file>\n"
	     "   -b <val> : contours around values != <val> [default: 1.0]\n"
	     "  -d <name> : use dataset <name> in the input files (default: first dataset)\n"
	     "           -- you can also specify a dataset via <filename>:<name>\n"
	  );
}

/* given an fname of the form <filename>:<data_name>, return a pointer
   to a newly-allocated string containing <filename>, and point data_name
   to the position of <data_name> in fname.  The user must free() the
   <filename> string. */
static char *split_fname(char *fname, char **data_name)
{
     int fname_len;
     char *colon, *filename;

     fname_len = strlen(fname);
     colon = strchr(fname, ':');
     if (colon) {
          int colon_len = strlen(colon);
          filename = (char*) malloc(sizeof(char) * (fname_len-colon_len+1));
          CHECK(filename, "out of memory");
          strncpy(filename, fname, fname_len-colon_len+1);
	  filename[fname_len-colon_len] = 0;
          *data_name = colon + 1;
     }
     else { /* treat as if ":" were at the end of fname */
          filename = (char*) malloc(sizeof(char) * (fname_len + 1));
          CHECK(filename, "out of memory");
          strcpy(filename, fname);
          *data_name = fname + fname_len;
     }
     return filename;
}

int main(int argc, char **argv)
{
     arrayh5 a, contour_data;
     char *png_fname = NULL, *contour_fname = NULL, *data_name = NULL;
     int *mask = NULL;
     double background_value = 1.0;
     double min = 0, max = 0;
     int min_set = 0, max_set = 0;
     extern char *optarg;
     extern int optind;
     int c;
     int slicedim = 2, islice = 0;
     int err;
     int nx, ny;
     colormap_t colormap = GRAYSCALE;
     int verbose = 0;
     int transpose = 0;
     double scalex = 1.0, scaley = 1.0;
     int invert = 0;
     double skew = 0.0;
     int ifile;

     while ((c = getopt(argc, argv, "ho:x:y:z:cm:M:C:b:d:vX:Y:Trs:V")) != -1)
	  switch (c) {
	      case 'h':
		   usage(stdout);
		   return EXIT_SUCCESS;
	      case 'V':
		   printf("h5topng " VERSION " by Steven G. Johnson\n" 
			  COPYRIGHT);
		   return EXIT_SUCCESS;
	      case 'v':
		   verbose = 1;
		   break;
	      case 'T':
		   transpose = 1;
		   break;
	      case 'r':
		   invert = 1;
		   break;
	      case 'o':
		   png_fname = (char*) malloc(sizeof(char) *
					      (strlen(optarg) + 1));
		   CHECK(png_fname, "out of memory");
		   strcpy(png_fname, optarg);
		   break;
	      case 'd':
		   data_name = (char*) malloc(sizeof(char) *
					      (strlen(optarg) + 1));
		   CHECK(data_name, "out of memory");
		   strcpy(data_name, optarg);
		   break;		   
	      case 'C':
		   contour_fname = (char*) malloc(sizeof(char) *
						  (strlen(optarg) + 1));
		   CHECK(contour_fname, "out of memory");
		   strcpy(contour_fname, optarg);
		   break;
	      case 'x':
		   islice = atoi(optarg);
		   slicedim = 0;
		   break;
	      case 'y':
		   islice = atoi(optarg);
		   slicedim = 1;
		   break;
	      case 'z':
		   islice = atoi(optarg);
		   slicedim = 2;
		   break;
	      case 'c':
		   colormap = BLUE_WHITE_RED;
		   break;
	      case 'm':
		   min = atof(optarg);
		   min_set = 1;
		   break;
	      case 'M':
		   max = atof(optarg);
		   max_set = 1;
		   break;
	      case 'b':
		   background_value = atof(optarg);
		   break;
	      case 'X':
		   scalex = atof(optarg);
		   break;
	      case 'Y':
		   scaley = atof(optarg);
		   break;
	      case 's':
		   skew = atof(optarg) * 3.14159265358979323846 / 180.0;
		   break;
	      default:
		   fprintf(stderr, "Invalid argument -%c\n", c);
		   usage(stderr);
		   return EXIT_FAILURE;
	  }
     if (optind == argc) {  /* no parameters left */
	  usage(stderr);
	  return EXIT_FAILURE;
     }

     if (contour_fname) {
	  int cnx, cny;

	  if (verbose)
	       printf("reading contour data from \"%s\".\n", contour_fname);
	  err = arrayh5_read(&contour_data, contour_fname, NULL, 
			     slicedim, islice);
	  CHECK(!err, arrayh5_read_strerror[err]);
	  CHECK(contour_data.rank >= 1,
		"data must have at least one dimension");
	  
	  cnx = contour_data.dims[0];
	  cny = contour_data.rank >= 2 ? contour_data.dims[1] : 1;
	  
	  mask = (int *) malloc(sizeof(int) * cnx * cny);
	  CHECK(mask, "out of memory");
	  compute_outlinemask(cnx,cny, contour_data.data,
			      mask, background_value);
     }
     
     for (ifile = optind; ifile < argc; ++ifile) {
          char *dname, *h5_fname;
          h5_fname = split_fname(argv[ifile], &dname);
          if (dname[0])
               data_name = dname;

	  if (verbose)
	       printf("reading from \"%s\", slice at %d in %c dimension.\n",
		      h5_fname, islice, slicedim + 'x');
	  
	  err = arrayh5_read(&a, h5_fname, data_name, slicedim, islice);
	  CHECK(!err, arrayh5_read_strerror[err]);
	  CHECK(a.rank >= 1, "data must have at least one dimension");
	  
	  CHECK(!contour_fname || arrayh5_conformant(a, contour_data),
		"contour data must be conformant with source data");
	  
	  if (!png_fname) {
	       png_fname = (char *) malloc(sizeof(char) * 
					   (strlen(h5_fname) + 5));
	       strcpy(png_fname, h5_fname);
	       
	       /* remove ".h5" from filename: */
	       if (strlen(png_fname) >= strlen(".h5") &&
		   !strcmp(png_fname + strlen(png_fname) - strlen(".h5"),
			   ".h5"))
		    png_fname[strlen(png_fname) - strlen(".h5")] = 0;
	       
	       /* add ".png": */
	       strcat(png_fname, ".png");
	  }
	  
	  {
	       double a_min, a_max;
	       arrayh5_getrange(a, &a_min, &a_max);
	       if (verbose)
		    printf("data ranges from %g to %g.\n", a_min, a_max);
	       if (!min_set)
		    min = a_min;
	       if (!max_set)
		    max = a_max;
	       if (min > max) {
		    invert = !invert;
		    a_min = min;
		    min = max;
		    max = a_min;
	       }
	  }
	  
	  nx = a.dims[0];
	  ny = a.rank < 2 ? 1 : a.dims[1];
	  
	  if (verbose)
	       printf("writing \"%s\" from %dx%d input data.\n",
		      png_fname, nx, ny);
	  
	  writepng(png_fname, nx, ny, transpose, skew,
		   scaley, scalex, a.data, mask, min, max, invert, colormap);

	  arrayh5_destroy(a);
	  free(png_fname); png_fname = NULL;
	  free(h5_fname);
     }
     if (contour_fname)
	  arrayh5_destroy(contour_data);
     free(mask);
     free(contour_fname);

     return EXIT_SUCCESS;
}
